Exo 2 solution

GregoireHENRY opened the pull request:
Implementation of encryption and decryption of Caesar encryptions with custom key

GregoireHENRY committed:
Correct decipher call

+ was sending plain text to be ciphered again..

GregoireHENRY committed:
Cipher plain text with custom caesar key

+ implementation do_cipher

GregoireHENRY committed:
Can decipher caesar encrypted messages

+ implementation of decipher

GregoireHENRY committed:
Test improvement suggestion

pymaldebaran commented on Exercice_2_-_Le_chiffre_de_Cesar/caesar.py:
```diff
@@ -107,7 +107,7 @@ def main():
 
     print("Let's decipher (will it work?!)")
-    DECIPHER_TEXT = do_cipher(PLAIN_TEXT, KEY)
+    DECIPHER_TEXT = do_decipher(CIPHER_TEXT, KEY)
```
Ça a été corrigé dans le main. Tu peux récupérer la bonne version en mergeant le main vers ta branche.

Il y a plusieurs façon de le faire mais la plus simple (sans faire de `rebase` quoi...) c'est de faire :

```shell
git checkout greg # tu vas sur ta branche
git pull # tu vas récupère les dernières donnée du serveur git
git merge main # tu merge les dernière modif du main dans ta branche
```
Le truc c'est que si tu as plusieurs branches, une pour chacun de tes PR par exemple... et bien tu dois faire pareil pour ces branches là. D'où l'intéret de ne pas avoir trop de branche de branche de branche de branche sinon la propagation des hotfix comme celui-là devient très chiante.

pymaldebaran commented on Exercice_2_-_Le_chiffre_de_Cesar/caesar.py:
```diff
@@ -83,8 +83,12 @@
-    return "Il faudrait retourner un vrai truc ;)"
+    return "".join(
+        [
+            " " if letter == " " else chr((ord(letter.upper()) - 65 - key) % 26 + 65)
```
Il faut mettre `" "` dans une constante

pymaldebaran replied:
D'où sort ce 65 venu de nulle part ? ça devrait être calculée. Par rétroingénierie je suppose que c'est `ord("A")`
Idem pour le 26 :point_right: à mettre dans une constante et à calculer `ALPHABET_LEN =ord("Z") - ord("A") + 1`

pymaldebaran reviewed (changes requested):
Globalement ta solution me semble pas hyper lisible... et est hyper lente sur de grosse chaîne je pense. Tu ferais mieux d'utiliser `str.translate` et `str.maketrans` pour faire ta transformation de chaîne plutôt que de tout faire à la main.

pymaldebaran reviewed (changes requested):
Outre le fait que je fait les même remarques que pour le `do_cipher` tu aurais pu remarquer qu'il est inutile d'écrire un corps à `do_decipher` car celui de `do_cipher`... regarde bien quel est la seule différence entre cipher et decipher : le signe de la clé !

pymaldebaran commented on Exercice_2_-_Le_chiffre_de_Cesar/caesar.py:
```diff
@@ -121,6 +121,9 @@ def main():
 
     SUCCESS = PLAIN_TEXT.upper() == DECIPHER_TEXT
     print("Did it worked?", "OK :)" if SUCCESS else "Nope :(")
+    # test is not enough, should also check CIPHER_TEXT because encryption can
```
Certes le test n'est pas parfait... mais suffisant pour commencer. (On abordera la notion de test à couverture totale plus loin).

C'est à mettre sur la TODO list donc dans un beau fichier CHANGELOG.md :smile_cat:

GregoireHENRY committed:
String rotation implementation

+ to prepare cipher alphabet from alphabet and caesar key in order to create translation table

GregoireHENRY committed:
Changed encryption function to use translate table

+ import string
+ use shift to create cipher alphabet from key
+ use str.maketrans to create translation table
+ use str.translate to apply translation table
+ return upper cased cipher text

GregoireHENRY committed:
Changed decryption to be symmetric to encryption

GregoireHENRY committed:
Inline comments moved to CHANGELOD.md

GregoireHENRY committed:
Update CHANGELOG.md

pymaldebaran commented on Exercice_2_-_Le_chiffre_de_Cesar/caesar.py:
```diff
@@ -81,19 +81,27 @@
+    """
+    plain_text_upper = plain_text.upper()
+    # create intermediate cipher alphabet shifted with key
+    cipher_albaphet_upper = shift(string.ascii_uppercase, -key)
```
C'est une constante ce `cipher_alphabet_upper` :point_right: en majuscule

pymaldebaran commented on Exercice_2_-_Le_chiffre_de_Cesar/caesar.py:
```diff
@@ -81,19 +81,27 @@
+    # create intermediate cipher alphabet shifted with key
+    cipher_albaphet_upper = shift(string.ascii_uppercase, -key)
+    # translation table that map classic alphabet to cipher alphabet shifted with the key
+    translation_table = plain_text_upper.maketrans(string.ascii_uppercase, cipher_albaphet_upper)
```
Pareil `translation_table` est aussi une constante :point_right: majuscule svp

pymaldebaran commented on Exercice_2_-_Le_chiffre_de_Cesar/caesar.py:
```diff
@@ -81,19 +81,27 @@
+    plain_text_upper = plain_text.upper()
+    # create intermediate cipher alphabet shifted with key
+    cipher_albaphet_upper = shift(string.ascii_uppercase, -key)
+    # translation table that map classic alphabet to cipher alphabet shifted with the key
```
Tu devrait évoquer dans ton commentaire ce qui arrive au caratères qui ne sont pas dans ta table comme les espaces vu qu'ils sont évoqués dans le cahier des charges.

pymaldebaran commented on Exercice_2_-_Le_chiffre_de_Cesar/caesar.py:
```diff
@@ -105,12 +105,11 @@ def do_cipher(plain_text: str, key: int) -> str:
+    Decipher encrypted text from caesar encryption key.
+    Operations are symmetric to do_cipher, thus we call it with opposed key.
+    """
+    return do_cipher(cipher_text, -key)
```
Yesssssss

pymaldebaran commented on Exercice_2_-_Le_chiffre_de_Cesar/CHANGELOG.md:
```diff
@@ -0,0 +1,28 @@
+
+## 0.2 - 2020-11-23
+
+### .
```
C'est quoi ce truc avec juste un point ???

pymaldebaran commented on Exercice_2_-_Le_chiffre_de_Cesar/CHANGELOG.md:
```diff
@@ -0,0 +1,28 @@
+
+### .
+
++ .
```
Idem c'est quoi ce point ?

GregoireHENRY committed:
Changed case for constants

+ correctly capitalize constants in do_cipher

GregoireHENRY committed:
Added comment to explain space processing

GregoireHENRY committed:
Update CHANGELOG.md in v0.2.1
